### ISSUE TITLE
Using a brush to select a time range.

### ISSUE THREAD

**Issue by jhvanderven** [emeeks/semiotic#55]
I would like to create a component that can be used to select a timerange using semiotic. What I have in mind is a "line chart" with a time axis, to which I want to add a Brush. I cannot find any example on how to do that. I have this:
```
 const timeRangeData = [
  {
    id: "timeRange",
    color: "blue",
    data: [{ px: new Date(2016,6,1,0,0,0), py: 1 }, { px: new Date(), py: 1}]
  }
]
         <XYFrame
            xScaleType={scaleTime()}
            yExtent={[0,2]}
            xAccessor={'px'}
            yAccessor={'py'}
            lineDataAccessor={'data'}
            lineStyle={d => ({
              stroke: d.color,
              strokeWidth: "30px",
              opacity: 0.2,
            })}
            size={[1000,200]}
            lines={timeRangeData}
            margin={{ left: 50, top: 20, bottom: 70, right: 50 }}
            axes={[
              {
                orient: "bottom",
                tickFormat: d => moment(d).week()
              }
            ]}
          />
```

How do I add the interactivity?

**Comment by emeeks** [comment]
I have an example using the ORFrame brush but no minimal example for the XYFrame brush. I'll put one together because that's something everyone needs.

**Comment by emeeks** [comment]
Just posted a minimal example, let me know if you need any more explanation:
https://emeeks.github.io/semiotic/#/semiotic/linebrush

**Comment by jhvanderven** [comment]
Thanks for the quick reply. Unfortunately your example is not working at my end. If I comment out brush: "xBrush", the page renders the XYFrame. If I comment it in, I get an error, with the following stack trace:

```
Uncaught TypeError: Cannot read property 'selection' of null
    at SVGGElement.<anonymous> (InteractionLayer.js:149)
    at Dispatch.apply (dispatch.js:61)
    at customEvent (on.js:103)
    at Emitter.emit (brush.js:286)
    at Emitter.start (brush.js:274)
    at SVGGElement.<anonymous> (brush.js:222)
    at Selection../node_modules/semiotic/node_modules/d3-selection/src/selection/each.js.__webpack_exports__.a [as each] (each.js:5)
    at brush.move (brush.js:212)
    at Selection../node_modules/semiotic/node_modules/d3-selection/src/selection/call.js.__webpack_exports__.a [as call] (call.js:4)
    at Brush.createBrush (Brush.js:53)
    at Brush.componentDidMount (Brush.js:44)
    at ReactCompositeComponent.js:265
    at measureLifeCyclePerf (ReactCompositeComponent.js:75)
    at ReactCompositeComponent.js:264
    at CallbackQueue.notifyAll (CallbackQueue.js:76)
    at ReactReconcileTransaction.close (ReactReconcileTransaction.js:80)
    at ReactReconcileTransaction.closeAll (Transaction.js:206)
    at ReactReconcileTransaction.perform (Transaction.js:153)
    at batchedMountComponentIntoNode (ReactMount.js:126)
    at ReactDefaultBatchingStrategyTransaction.perform (Transaction.js:140)
    at Object.batchedUpdates (ReactDefaultBatchingStrategy.js:62)
    at Object.batchedUpdates (ReactUpdates.js:97)
    at Object._renderNewRootComponent (ReactMount.js:320)
    at Object._renderSubtreeIntoContainer (ReactMount.js:401)
    at Object.render (ReactMount.js:422)
    at Object../src/index.js (index.js:7)
    at __webpack_require__ (bootstrap c2061b57f03955bc38b1:669)
    at fn (bootstrap c2061b57f03955bc38b1:87)
    at Object.0 (registerServiceWorker.js:108)
    at __webpack_require__ (bootstrap c2061b57f03955bc38b1:669)
    at bootstrap c2061b57f03955bc38b1:715
    at bundle.js:719

```
Does the example require any props? I just add it to the page as
```
<LineBrush />
```

I have 1.1.4 of semiotic, react is still at 15.4.2

Here is my stripped version of your component:
```
import React from "react";
import { XYFrame } from "semiotic";
import { scaleTime } from "d3-scale";

const chartScale = scaleTime();
const lineStyle = {
  fill: "#007190",
  stroke: "#007190",
  strokeWidth: 1
};

var data = [
  {
    date: "1/2/2003",
    close: 14.8
  },
  {
    date: "12/30/1998",
    close: 14.07
  }
];

data.forEach(d => {
  d.date = new Date(d.date);
});

export default class LineBrush extends React.Component {
  constructor(props) {
    super(props);
    this.brushEnd = this.brushEnd.bind(this);
  }

  brushEnd(e) {
    console.log(e);
  }

  render() {
    return <XYFrame
        size={[700, 200]}
        lines={[{ label: "Apple Stock", coordinates: data }]}
        xAccessor={d => d.date}
        yAccessor="close"
        xScaleType={chartScale}
        lineStyle={lineStyle}
        axes={[
          {
            orient: "bottom",
            ticks: 6,
            tickFormat: d => d.getFullYear()
          }
        ]}
        margin={{ left: 40, top: 0, bottom: 50, right: 20 }}
        interaction={{
          end: this.brushEnd,
          brush: "xBrush",
          extent: [new Date("1/2/1997"), new Date("1/2/2003")]
        }}
      />
   };
}

```

**Comment by emeeks** [comment]
No, there is no <LineBrush> in semiotic, rather the example shows you how you can activate interactivity in XYFrame by setting the `interaction` property as in the example. I just named my example component `LineBrush`. I've avoided adding a bunch of syntactic sugar components to Semiotic (like <Scatterplot> or <LineChart>) so far. You would need to define the frame like that and the `brushEnd` function and presumably your own behavior in the `brushEnd` function.

**Comment by emeeks** [comment]
Here's that code in semiotic-examples and it seems to be working fine:
https://github.com/emeeks/semiotic-examples/blob/master/src/LineBrush.js

I think the error you're experiencing is in `d3-brush`. I believe the event in D3 has issues in some build environments:

https://github.com/d3/d3/issues/2733

As I understand Mike's position is that this is a failure in the toolchain and not an issue of D3 and I believe his response is that you should use a better build environment. If you or anyone else knows of or can find a workaround, I would be happy to implement it in Semiotic.

**Comment by jhvanderven** [comment]
This is not something I know enough off.
What kind of environment are you using?
I tried to find a create-react-app thingy using rollup, but what I found is not good enough.

**Comment by emeeks** [comment]
I'm sorry I know and share your frustration, we had to scrap a brush implementation in one of our older dashboards because of this bug. At some point I will either implement a native react brush or write my own because of this.

**Comment by jhvanderven** [comment]
I am very happy with your library and even with the brush as it is now. Just one more question: can I set the selected extent to something else than the whole extent when the state changes?

**Comment by emeeks** [comment]
You can initialize the brush with a set extent by sending an `extent` array to interaction (as is done with the LineBrush example above) but there is no way to procedurally update the extent currently. I'll file a feature request for that, since it's an important feature.

**Comment by brianretford** [comment]
I believe this is related to having a separate copy of d3 inside the node modules of semiotic itself, as well as having them included in node-package. If I `rm -rf node_modules/semiotic/node_modules` the problem does away

**Comment by emeeks** [comment]
You're saying you see `d3` installed in the node_modules of semiotic or a different `d3-*` version than what you've got in your main project?

**Comment by brianretford** [comment]
Yes, #1, and also in my main project, at least using create-react-app

**Comment by emeeks** [comment]
Could you share with me the repo somehow (privately or with the identifying bits obscured). That shouldn't be happening, there shouldn't be any dependencies anywhere to `d3` itself.

**Comment by brianretford** [comment]
It's pretty 'easy' to reproduce, though I think I have the reason it occurs wrong. Steps:
* Create a component that uses `MinimapXYFrame`
* Use `npx create-react-app broken`, and then add your custom component to it after doing `yarn add semiotic` -- broken
* Use `nwb create-react working`, add your component, `npm install --save semiotic` and everything works
* If you delete the nest node_modules in `broken`, everything magically works.

Comparing the lock files, it appears that the yarn-lock file has multiple version of some d3 components in it:
```
d3-collection@1, d3-collection@^1.0.4:
  version "1.0.4"
  resolved "https://registry.yarnpkg.com/d3-collection/-/d3-collection-1.0.4.tgz#342dfd12837c90974f33f1cc0a785aea570dcdc2"

d3-collection@1.0.1:
  version "1.0.1"
  resolved "https://registry.yarnpkg.com/d3-collection/-/d3-collection-1.0.1.tgz#5b5c5626f731122b600b107d71a08833fec049ad"
```

Things like this seem to be caused by yarn recording transitive dependencies and npm not, it's way past my ken to understand these things though.

yarn culprit:
```
d3-force@1.0.2:
  version "1.0.2"
  resolved "https://registry.yarnpkg.com/d3-force/-/d3-force-1.0.2.tgz#92dbedeacf9a2d3f6686450c8d6375f3c363af66"
  dependencies:
    d3-collection "1"
    d3-dispatch "1"
    d3-quadtree "1"
    d3-timer "1"
```

You can see that it depends on generic version `1` of a lot of components, which then result in yarn having a version war as seen above.  Wish I could say how to fix it.

**Comment by brianretford** [comment]
One possible solution might be to remove the very specific dependencies in semiotic, maybe only depend on `1.0`, or maybe even just `1`

**Comment by emeeks** [comment]
Yes you're right, I'll clean up dependencies on my next version. I know these exact version dependencies have caused some hassle with other libs.

**Comment by brianretford** [comment]
Cool, let me know if I can help. I tried 4 different d3 vis frameworks and
this one came out on top, especially for large datasets.

As a maintainer of https://github.com/plaidml/plaidml I totally understand
how complicated this stuff can get.

Really appreciate semiotic!

**Comment by emeeks** [comment]
I just published `1.9.6` that unfixes all the D3 deps, which also references a new version of `semiotic-mark` that does the same. Let me know if that helps or causes other issues.
